### ISSUE TITLE
Performance problem with TensorFlow training

### ISSUE THREAD

**Issue by fcacarminati** [tensorflow/tensorflow#19491]
Hello,
    We are running a not-very-complex 3D convolution problem had we have extremely poor performance. Here is the summary of our problem

Now the technical part.

I am running on an Haswell CPU in a Mac OS running High Sierra. 

Model Name:	MacBook Pro
Model Identifier:	MacBookPro11,5
Processor Name:	Intel Core i7
Processor Speed:	2.5 GHz
Number of Processors:	1
Total Number of Cores:	4
L2 Cache (per Core):	256 KB
L3 Cache:	6 MB
Memory:	16 GB

Tensorflow performance
---------------------------------

1.) Memory allocation

Memory allocation seems highly unoptimized. I see an allocation of ~80GB (78M allocations) out of which we are left with 37GB persistent (corresponding to 575k permanent allocations). The memory churn is enormous and this may affect very seriously performance. Most of those are very small allocation / deallocation which happen here

Eigen::NonBlockingThreadPoolTempl<tensorflow::thread::EigenEnvironment>::Schedule(std::__1::function<void ()>)

I have tried to run with tcmalloc from google hoping to improve memory allocation and handling. tcmalloc complains that there are the following “large allocation” from TF even before starting the epochs: (23+23+2+18+4+9+2) = 81GB of allocation even before starting the epoch’s. After that my disk is full of swap files and my machine dies.

Then I went back running with the Mac allocator, which surprisingly seems to be more robust. 

2.) Code performance

A careful VTune analysis performed by Sofia has identified Eigen as the major source of CPU consumption.  All the time is wasted simply in repacking (gemm_pack_rhs).

To look at the code I attempted to compile with -g, however the default compilation is with -g0 and I could not find yet a way to replace this default on bazel. I added -g3 that, according to the manual (and to a small test I have made) should override -g0. However the Mac Instrument (a poor relation of VTune on Mac) could not find out the source. The library should be _pywrap_tensorflow_internal.so. Then I went looking for the source and I found that gemm_pack_rhs::operator() is defined in the following files

./bazel-tensorflow/external/eigen_archive/Eigen/src/Core/products/GeneralBlockPanelKernel.h
./bazel-tensorflow/third_party/eigen3/unsupported/Eigen/CXX11/src/FixedPoint/MatMatProductAVX2.h
./third_party/eigen3/unsupported/Eigen/CXX11/src/FixedPoint/MatMatProductAVX2.h

The last two are identical. Putting good old printf’s I discovered we are calling the GeneralBlockPanelKernel.h version. The operator works with packets of size 8, which is fine for AVX2 (256) and float32 as we are using. However I am not sure that the compiler manages to vectorize this procedure. Indeed, most of the time is spent in line 559 of eigen_volume_patch.h. 

   for (int i = 0; i < PacketSize; ++i) {
      values[i] = coeff(index + i);
    }

The packet structure of the code is meant to have each packet treated as a unit. This for loop simply destroys all possibility of optimisation. There is a lot of room for optimisation in tensorflow before we get really serious about performance with a problem like ours. But who is going to pick up the tab? 

3.) MKL or not MKL. 

When bringing up this problem, I have been answered that tensorflow in Mac does not support the usage of MKL, and therefore, till then, my findings were not entirely relevant. MKL for Mac exists, however clang does not support OMP (or rather the default version of clang distributed with Mac does not have OMP support enabled). So the only way to compile tensorflow on the Mac with MKL was to change compiler. 

Unfortunately changing compiler with bazel on the Mac seems a very ambitious proposition. After posting to and perusing stackoverflow, bazel forum and tensorflow forum, I came to the following recipe

export BAZEL_USE_CPP_ONLY_TOOLCHAIN=1 
export CC=/path/to/compiler
bazel build […]

does indeed force Bazel to use a new compiler, however controlling the compiler switches is much more complicated. The two compiler flags -Wthread-safety and -Wself-assign, as well as the linker flag “-no-as-needed” and “-z” are incompatible with g++ linker. The CROSSTOOL.tpl are automatically generated during configuration. The only occurrences of (for instance) -Wself-assign in the TF code are in 

third_party/gpus/crosstool/CROSSTOOL_nvcc.tpl
third_party/toolchains/clang6/CROSSTOOL.tpl
third_party/toolchains/cpus/arm/CROSSTOOL.tpl

but even if I comment the lines:

-  compiler_flag: "-Wthread-safety"
-  compiler_flag: "-Wself-assign"
+#  compiler_flag: "-Wthread-safety"
+#  compiler_flag: "-Wself-assign”

in all three of them “something” creates a CROSSTOOL.tpl with these flags in. The hack I am trying now is to configure and then edit the file

./bazel-tensorflow/external/local_config_cc/cc_wrapper.sh

adding the following line

/sw/bin/gcc-fsf-7 `echo "$@" | sed -e 's/-Wself-assign//' | sed -e 's/-Wthread-safety//' | sed -e 's/-Wl,-no-as-needed//' | sed -e 's/-Wl,-z,relro,-z,now//‘`

which is a very poor hack. 

With this I could build a version of tensorflow using the Mac MKL, but to no avail. Performance is still abysmal with the same bottleneck. 

Thanks for reading up to here...

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by fcacarminati** [comment]
Hello,
    I have no idea how to access the template. Sorry for my ignorance. Can you help? Best,

**Comment by fcacarminati** [comment]
Forgot the keras file

[keras.json.txt](https://github.com/tensorflow/tensorflow/files/2044780/keras.json.txt)

**Comment by fcacarminati** [comment]
A small update on this matter. When we make the filters multiple of 16 and the dataset size multiple of 32, the training speed improves by a factor 30. Apparently when these two conditions are met, the data are repacked by MKL, which is much faster. Otherwise some slower functions are used, and the difference is massive.

**Comment by vpirogov** [comment]
@fcacarminati, Filters being a multiple of 16 is recommended to get the best performance on systems with Intel AVX512 instruction set. For systems with Intel AVX2 support (like your Macbook) multiple of 8 is enough. There's no restrictions related to the dataset size or the minibatch size, though the larger minibatch will results in better performance.

**Comment by angersson** [comment]
I'm not sure I understand the core of this issue. Can you explain what kind of response you are looking for here?

**Comment by fcacarminati** [comment]
Thanks for your comment. My point is that under some conditions (i.e. filters not being multiple of 8 or 16), the training performance is sub-optimal and half of the time is spent repacking tensors.

**Comment by fcacarminati** [comment]
Thanks a lot… still the performance should be better, and I am pointing out where I think it could be improved. Best,

Federico Carminati
Chief Innovation Officer
CERN openlab
1211 Geneva 23
Switzerland
Tel: +41 22 76 74959
Fax: +41 22 76 68505
Mobile: +41 75 411 4843

On Jun 1, 2018, at 19:12, Vadim Pirogov <notifications@github.com<mailto:notifications@github.com>> wrote:


@fcacarminati<https://github.com/fcacarminati>, Filters being a multiple of 16 is recommended to get the best performance on systems with Intel AVX512 instruction set. For systems with Intel AVX2 support (like your Macbook) multiple of 8 is enough. There's no restrictions related to the dataset size or the minibatch size, though the larger minibatch will results in better performance.

**Comment by angersson** [comment]
@tfboyd Can you take a look at this?

**Comment by tfboyd** [comment]
@tatianashp    Assigning to the CPU performance lead.  No low hanging fruit for me to grab.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatianashp: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatianashp: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatianashp: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by fcacarminati** [comment]
Actually I would really like an answer... thanks!

**Comment by tatianashp** [comment]
@fcacarminati Thank you for pointing out a performance problem. I apologize that this issue fell through the cracks. If I understand your request correctly, you are asking if we could improve performance of
the loop in line 559 of eigen_volume_patch.h.
    for (int i = 0; i < PacketSize; ++i) {
        values[i] = coeff(index + i);
    }
Is this correct?

/cc @ezhulenev @rmlarsen

**Comment by fcacarminati** [comment]
Thanks for looking into my problem. In my problem 50% of the time is spent in this line. The packet should be executed in a single vector instruction, ideally. There it is executed in a loop calling other procedures. This kills optimisation, as far as I understand. Regards,

Federico Carminati
Chief Innovation Officer
CERN openlab
1211 Geneva 23
Switzerland
Tel: +41 22 76 74959
Fax: +41 22 76 68505
Mobile: +41 75 411 4843

On Jul 20, 2018, at 20:07, Tatiana Shpeisman <notifications@github.com<mailto:notifications@github.com>> wrote:


@fcacarminati<https://github.com/fcacarminati> Thank you for pointing out a performance problem. I apologize that this issue fell through the cracks. If I understand your request correctly, you are asking if we could improve performance of
the loop in line 559 of eigen_volume_patch.h.
for (int i = 0; i < PacketSize; ++i) {
values[i] = coeff(index + i);
}
Is this correct?

/cc @ezhulenev<https://github.com/ezhulenev> @rmlarsen<https://github.com/rmlarsen>

**Comment by tatianashp** [comment]
@ezhulenev Could you take a look at what's going on here?

**Comment by ezhulenev** [comment]
@fcacarminati Unfortunately in your case the volume patch data is scattered across all of the input buffer, and there is no contiguous blocks of 8+ scalars, so TensorVolumePatchOp packet access has to construct packet by loading scalars one by one.

One option to make it faster, is to create custom LHS/RHS packer, similar to what you can find in eigen_spatial_convolutions, I'll take a look on it, it does seem like relatively simple thing. That change made Conv2D ~2x faster, so I expect to see somewhat similar results for Conv3D.

**Comment by fcacarminati** [comment]
[just back from vacation] Thanks a lot for your answer. I am really glad you acknowledged this issue. We are really dependent on the training performance for our studies and any improvement could be a big help. Please keep us informed about your progress. Best regards,

**Comment by tensorflowbutler** [comment]
Nagging Assignees @tatianashp, @ezhulenev: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @tatianashp, @ezhulenev: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by ezhulenev** [comment]
I'm working on it right now, have 10x-125x improvements for the Conv3D and Conv3DBackward, should land in master soon.

**Comment by fcacarminati** [comment]
Hello, 
   this is fantastic news and we cannot wait to test it. If you want, you can give us the code and we will try it out and let you know. Thanks again for your help. Best,

**Comment by ezhulenev** [comment]
@fcacarminati I've submitted couple of changes that should make Conv3D performance much better (I'd expecte something like ~5x). One of them is custom kernels for backprop input and filter: https://github.com/tensorflow/tensorflow/commit/e183b8d0328d7398cb6ffc530d1ae8fdbd2111c0

These kernels make less smaller allocations, instead they allocate quite large temporary buffers, so you might see increased peak memory usage. It's possible to fallback on original Eigen kernels using `kernel_label_map` (see https://github.com/tensorflow/tensorflow/blob/8cb0558da924e891aa1bb5d79a6c0c846301e4eb/tensorflow/python/framework/ops.py#L3311), old kernels registered with a "eigen_tensor" label.

I'd be super interested to know how much this helps in your specific case.

**Comment by fcacarminati** [comment]
Thanks a lot! I have tested immediately your changes and I have a 2.5x speedup on my Mac (haswell). We will make more conclusive tests on a GPU system, but this is already very very good. More news soon. Thanks again for your help, this makes the difference already for us,

**Comment by ezhulenev** [comment]
No, these changes are for CPU only, GPU kernels are the same cudnn base code as before.

Can you upload the graph def that you are executing on CPU, TF CPU convolutions require NDHWC* data format, and from your code I see that Keras is using channels_first. I guess there might be some unnecessary layout conversions. Also I don't have a good representative Conv3D dimensions for benchmarks: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/eigen_benchmark_cpu_test.cc#L403

*) Channels first works best for GPU (aka NCDHW).

**Comment by tensorflowbutler** [comment]
Nagging Assignees @tatianashp, @ezhulenev: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by fcacarminati** [comment]
> @fcacarminati I've submitted couple of changes that should make Conv3D performance much better (I'd expecte something like ~5x). One of them is custom kernels for backprop input and filter: [e183b8d](https://github.com/tensorflow/tensorflow/commit/e183b8d0328d7398cb6ffc530d1ae8fdbd2111c0)
> 
> These kernels make less smaller allocations, instead they allocate quite large temporary buffers, so you might see increased peak memory usage. It's possible to fallback on original Eigen kernels using `kernel_label_map` (see
> 
> [tensorflow/tensorflow/python/framework/ops.py](https://github.com/tensorflow/tensorflow/blob/8cb0558da924e891aa1bb5d79a6c0c846301e4eb/tensorflow/python/framework/ops.py#L3311)
> Line 3311 in [8cb0558](/tensorflow/tensorflow/commit/8cb0558da924e891aa1bb5d79a6c0c846301e4eb)
>  def _kernel_label_map(self, op_to_kernel_label_map): 
> ), old kernels registered with a "eigen_tensor" label.
> I'd be super interested to know how much this helps in your specific case.

I have a factor 3 improvement! And I have a factor 2.5 with the 3D patch changes. All in all this would give me a fantastic factor close to 10. When can we expect these two changes to get into the same branch? We are *really* depending on performance here and the work seems to be mostly done. One more little effort please :-) Best regards

**Comment by tensorflowbutler** [comment]
Nagging Assignees @tatianashp, @ezhulenev: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @tatianashp, @ezhulenev: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @tatianashp, @ezhulenev: It has been 45 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @tatianashp, @ezhulenev: It has been 63 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by rmothukuru** [comment]
@fcacarminati,
Sorry for the delayed response. 

As there were many upgrades in **`Tensorflow, CUDA and CuDNN`** versions since past 3 years, can you please confirm if your issue still persist? Thanks!
